### PR TITLE
Refrain from passing `encoding` to the SQL engine in SQLAlchemy v2

### PR DESCRIPTION
**Pull request by Dev-iL**
This PR fixes 1 error and 1 warning related to SQLAlchemy v2 ("SQLA2"), as well as an unrelated documentation hyperlink issue.
See motivation for changes here: https://github.com/apache/airflow/issues/28723#issuecomment-1951118883
related: #28723 

Note that while the added unit tests pass on SQLA1, one of them fails with SQLA2, and will require additional future modifications before that version could be adopted:
```none
ERROR test_orm_event_handlers.py::test_setup_event_handlers_no_warnings - sqlalchemy.exc.ArgumentError: Type annotation for "TaskInstance.dag_model" can't be correctly interpreted for Annotated Declarative Table form.  ORM annotations should normally make use of the ``Mapped[]`` generic ...
```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by Dev-iL** [comment]
@potiuk Pinging as promised.

File: tests/utils/test_orm_event_handlers.py
Comment (Taragolis) [inline tests/utils/test_orm_event_handlers.py]:
```suggestion
```

File: tests/utils/test_orm_event_handlers.py
Comment (Taragolis) [inline tests/utils/test_orm_event_handlers.py]: You could just setup to catch this kind of warnings and raise an error

```suggestion
        warnings.simplefilter("error", category=SADeprecationWarning)
        try:
            setup_event_handlers(engine=None)
        except InvalidRequestError:
            # We are only interested in the warning that may be issued at the beginning of the method.
            pass
```

File: tests/utils/test_orm_event_handlers.py
Comment (Taragolis) [inline tests/utils/test_orm_event_handlers.py]: If you setup warnings filter, then you do not need this code


```suggestion
```

File: airflow/settings.py
Comment (Taragolis) [inline airflow/settings.py]: We don't have a discussion yet about whether or not we should use `__version__` or `importlib.metadata`.
@potiuk any thoughts? I'm personally have no objections to use it for now, because SQLachemy maintainers do not announced about deprecate/remove `__version__`

File: tests/utils/test_orm_event_handlers.py
Comment (Dev-iL) [reply tests/utils/test_orm_event_handlers.py]: You're absolutely right, I didn't mean to include that part. Will remove.

File: airflow/settings.py
Comment (Dev-iL) [reply airflow/settings.py]: I originally did this (in a new util file dedicated to dependency version checks):
```python
from packaging.version import Version, parse as parse_version
from sqlalchemy import __version__ as sqlalchemy_version


def is_sqlalchemy_v1() -> bool:
    """Check if SQLAlchemy v1 is installed."""
    installed: Version = parse_version(sqlalchemy_version)
    return installed.release[0] == 1
```

But then saw the `PENDULUM3` example and amended my code. I don't mind changing it to anything else.

File: tests/utils/test_orm_event_handlers.py
Comment (Taragolis) [reply tests/utils/test_orm_event_handlers.py]: Sometimes this part remaining. That why is review exists.

BTW, just wondering why this is required? IDE can't figure out that this module contain tests?

Comment (Dev-iL) [reply tests/utils/test_orm_event_handlers.py]: The reason for adding it was so I could debug the test in the IDE, otherwise I have to run the test using the CLI, which is less useful.

File: airflow/settings.py
Comment (Taragolis) [reply airflow/settings.py]: No I mean a bit different, e.g. get from distribution package, but we do not discuss preferable way

```python
import sys
if sys.version_info >= (3, 9):
    from importlib.metadata import version
else:
    from importlib_metadata import version

SQL_ALCHEMY_V1 = version("sqlalchemy").startswith("1")
```


> But then saw the PENDULUM3 example and amended my code.

Pendulum3 I guess exists only in tests

Comment (Dev-iL) [reply airflow/settings.py]: 1. [Not only in tests](https://github.com/apache/airflow/blob/016892b7d6bc654ac3ff7e4f8f9f7231f835bed7/airflow/utils/timezone.py#L32).
2. Importlib metadata was [added in py3.8,](https://docs.python.org/3/library/importlib.metadata.html) which is the minimum supported in airflow anyway, so no need to do any fancy checking.

Comment (Taragolis) [reply airflow/settings.py]: > [Not only in tests](https://github.com/apache/airflow/blob/016892b7d6bc654ac3ff7e4f8f9f7231f835bed7/airflow/utils/timezone.py#L32).

Oh... yeah... that reminds me that we decide to not to use try except. But most possible pendulum 2 would be dropped into the Airflow 2.9 or 2.10

> Importlib metadata was [added in py3.8,](https://docs.python.org/3/library/importlib.metadata.html) which is the minimum supported in airflow anyway, so no need to do any fancy checking.

Importlib mandatory since 3.10, in cpython, I guess, most of the things implemented in 3.9. And something more, airflow read entire environment metadata during the providers loading, so this not a big deal even cache all packages installed into the environment. 

Anyway this one just a suggestion, no objections at this point. It could be changed in the future, we still not fully support SA2.0, so we have plenty a time. The point here, that some of package stop provide `__version__` (e.g. Flask and everything related to it) or newer provide it.

Comment (Dev-iL) [reply airflow/settings.py]: I ended up using try/except after trying, without success, to mock an engine (using MagicMock, and sqlalchemy.engine.mock). I couldn't get past the event registration which throws the TypeError I ended up catching. It was either that or get rid if the test entirely. If you (or anyone else) has an idea how to write this test w/o try/except - I'll be happy to hear.

BTW, what's wrong with try/except? I thought EAFP was the preferred way to write python code.

Comment (Taragolis) [reply airflow/settings.py]: > I thought EAFP was the preferred way to write python code.

Nothing wrong, did I tell that it wrong? In general everything depends on the situation, depend on the cases. There was more hacky case rather than check the version

Comment (Taragolis) [reply airflow/settings.py]: In this case it is fine. 

BTW, could you rebase your branch? It should fix failing tests.

**Review by Taragolis — APPROVED** [review]

File: airflow/settings.py
Comment (potiuk) [inline airflow/settings.py]: Shoudln't that be "1." ? (It's purely hypotethical of course, but what happens when we have SqlAlchemy 10? :D

**Review by potiuk — APPROVED** [review]
Appproved - with a potential NIT

File: airflow/settings.py
Comment (Taragolis) [reply airflow/settings.py]: I'm pretty sure I hear that case before, let me think...

Windows 95 -> ... -> Windows 8 -> Windows 10

Comment (potiuk) [reply airflow/settings.py]: Window ME was that one betwen 95 and 8. And well this is what happened :D

image

**Review by dirrao — APPROVED** [review]

File: airflow/settings.py
Comment (Dev-iL) [reply airflow/settings.py]: 48c

Sure, will fix this when rebasing.

Comment (uranusjr) [reply airflow/settings.py]: Since we have `packaging` available and this is a one-time calculation, I’d try to properly parse the version with `packaging.version.Version` to be the most technically correct. And maybe read the actual package version with `importlib.metadata` instead.

But even more ideally, is there a way to actually detect whether the `encoding` parameter is supported or not? Feature detection is always better than agent detection if possible (this is borrowed from web frontend but also applies well here).

File: airflow/utils/orm_event_handlers.py
Comment (uranusjr) [inline airflow/utils/orm_event_handlers.py]: Maybe it is better to create a compat module for SQLALchemy to handle this? Or maybe this sort of compat code can go into `airflow.utils.sqlalchemy` instead to avoid cluttering the code base.

File: airflow/settings.py
Comment (Dev-iL) [reply airflow/settings.py]: @uranusjr Thank you for the comment. This was also raised in one of the other threads: https://github.com/apache/airflow/pull/37545#discussion_r1494649517

I don't mind adding a check using `importlib` and `packaging`, but I think it shouldn't go to the sqlalchemy util, since as mentioned, we have another example of checking the dependency version similarly (pendulum). I say these checks should go into a dedicated "dependency versions" (or somesuch) utility. Thoughts?

Comment (uranusjr) [reply airflow/settings.py]: I don’t like putting all the checks (of different things) in one module, but maybe putting SQLA checks in the utils.sqlalchemy submodule?

File: airflow/utils/orm_event_handlers.py
Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: I'm not sure how to create such a module + it might be out of scope for this PR. I did move the check as suggested.

**Comment by uranusjr** [comment]
Eh this is not what I suggested, I want the SQLA v1 check to go into `airflow.utils.sqlalchemy`, and the others kept as-is.

File: airflow/utils/orm_event_handlers.py
Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: This is easy - just wrap checking the sqlalchemy check in a function and move the fuction to a new `sqlalchemy.py` file in  `airflow.utils` folder - and I agree with @uranusjr - having it a separate util is a good idea. 

Se for example some checks in `airflow.utils.pydantic` module where we are doing very similar check - whether pydantic 1 or 2 or none is installed.

**Review by potiuk — CHANGES_REQUESTED** [review]
Agree with @uranusjr -> we should move it to `airflow.uitls.sqlalchemy`

File: airflow/utils/orm_event_handlers.py
Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: @potiuk Do you not want to avoid code duplication? Why is it preferable to copy this entire block while changing one line over refactoring the version checking to a separate utility where the library and target version are inputs, then invoke from everywhere as needed?
```python
def is_pydantic_2_installed() -> bool:
    import sys

    from packaging.version import Version

    if sys.version_info >= (3, 9):
        from importlib.metadata import distribution
    else:
        from importlib_metadata import distribution
    try:
        return Version(distribution("pydantic").version) >= Version("2.0.0")
    except ImportError:
        return False
```

**Comment by Dev-iL** [comment]
@uranusjr @potiuk I moved the methods as suggested. Please resolve threads and approve if acceptable.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
LGTM. @uranusjr ?

**Comment by Dev-iL** [comment]
Can someone please explain the build failure to me? I don't know what to do about it.

**Comment by Taragolis** [comment]
This error is expected, regression of https://github.com/apache/airflow/pull/37524
Every PR now have this error, except the PRs which do not require to run `LatestBoto` ci tests

File: airflow/utils/orm_event_handlers.py
Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: Not sure what you are suggesting ?

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: I did not suggest to copy the block (if that's what you think) - I was showing that we already have similar approach for Pydantic check (with more caveats because we also do fake imports there to make our code works without Pydantic).

Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: I suggest having a utility that's not specific to pydantic, pendulum sqlalchemy etc., something like:
```python
# import packaging, metadata, version, ...

def _get_lib_major_version(lib_name: str) -> int:
    ver: Version = parse_version(version(lib_name))
    return ver.major

def is_installed_lib_major_version(libname: str, version: int) -> bool:
    return _get_lib_major_version(lib_name) == version

# Then, in a TBD location:
def is_sqlalchemy_v1() -> bool:
    return is_installed_lib_major_version("sqlalchemy", 1)
```

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: Feel absolutely free to make a refactor and propose a PR . This is how things work here. Someone sees an opportunity to improve things and proposes PRs :) , Maybe also find other places where such thing might be used (BTW. I think in this PR and in proposed solution you should merge the first two mehtods, this will be way shorter and more readable.

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]:
```python
def is_installed_lib_major_version(libname: str, version: int) -> bool:
    return parse_version(version(lib_name)).major == version
```

Or possibly even avoid  common another common utility but simply use that `one liner` everywhere (if it works also for various python versions - there is a good reason why we were checking for importlib in the previous solution)

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: BTW, What's the difference between using `distribution` and `version` (and how it works for different Python versions?) did you check it?

Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: I did include the first of them. Can you suggest a suitable "general" place for them? Should I add another file to the utils folder?

It's not exactly a one-liner since it comes with a bunch of imports.

I did not compare `distribution` with `version`, mainly because the latter seemed to do what I need.

Re `importlib` - it seems to me that the verbose version is for py<3.8 (`importlib_metadata` is a backport). I didn't understand @Taragolis' [explanation](https://github.com/apache/airflow/pull/37545#discussion_r1494959852) why this is not the case.

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: Ah yeah - seems we can use importlib.metadata version directly - the importlib_metadata imports are not needed as we are not using an features that would not be available in Python 3.8  - so yes, feel free to simplify all that (probably in a separate PR).

Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: I really want to do something like this, but don't know where to put it. The main reason for not wanting to append it to an existing file is because I want it to be as standalone as possible, so that it cannot ever cause circular imports.

```python
# e.g. airflow/utils/dependency_versions.py
from importlib.metadata import version as get_version_string

from packaging.version import parse as parse_version

def is_installed_lib_major_version(lib_name: str, major_version: int) -> bool:
    """Check if an installed library has the specified major version."""   
    return parse_version(get_version_string(lib_name)).major == major_version
```
Then wherever this is needed will either have a helper wrapper or be used directly
```python
# airflow/utils/sqlalchemy.py, or airflow/settings.py
from airflow.utils.dependency_versions import is_installed_lib_major_version

# OPTION 1:
def is_sqlalchemy_v1() -> bool:
    return is_installed_lib_major_version("sqlalchemy", 1)

...
if is_sqlalchemy_v1():
    # do something

# OPTION 2:
if is_installed_lib_major_version("sqlalchemy", 1):
    # do something
```

----

BTW, why is `airflow.utils.orm_event_handlers` a separate file and not part of `airflow.utils.sqlalchemy`?

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: create a separate file -propose a name, or find a good one. Generally we expect to autonomously propose somethign by looking at the code and doing things similarly. If it looks good, we will accept it, if not, we will suggest changes.

> BTW, why is airflow.utils.orm_event_handlers a separate file and not part of airflow.utils.sqlalchemy?

Because it's author chose it to be in a sparate package I guess?  There are no strict rules for that  and is (as many naming decisions) subject to individual person and reviewer. But if you have an idea and need to introduce some rules for naming that will be easy to answer question `why` then  I think you should propose it.

Gennerally - it's not a science , it's art and that art depends very much with artist you interact with. You have to accept ambiguities, some level of uncertainty and show initiative on your own (and accept the fact that commiters might have different (sometimes between them) opinions, it's also up to you to figure out how to get to consensus for names etc.

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: `Cache invalidation, naming things and off-by one mistakes are the two most difficult things in comuputer science`

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: BTW.  Speaking of making decisions and changing them.  After looking there I actually think that it is really not worth to make a common module for that. Since REALLY you can reduce it to one-liner, better is to use that one-liner everywhere, rather than extract methods. 

It's a rookie mistake to think that DRY is always best. It's not. sometimes inlining similar code is better for readability. performance and  reduced coupling ( in this case coupling causes too many complications - spmetims we will want to check for major version, sometimes for >= , sometimes maybe other check, so it's actually better to inline the checks and leave the freedon on what condition we want to check. Sometimes we will want to check this:

```python
parse_version(get_version_string('package')) >= Version('2.0.0`) 
```

but sometimes

```python
parse_version(get_version_string('package')).major = 1
```

So I think it makes very little sense to extract it to common, code, better is to make sure that similar patterns are used everywhere and coupling reduced between those different modules (not mentioning circular dependencies).

Comment (Dev-iL) [reply airflow/utils/orm_event_handlers.py]: Thank you for the insights @potiuk.

I honestly don't know what to do about this PR anymore, and it is frustrating. It was supposed to be a very minor change that got blown out of proportion. The PR seems to be "good enough" according to the reviewers, and yet it's stuck.

What do I need to do to get it merged? Keep rebasing it and hope for a slot where nothing else merges in the meantime? Resolve all threads unilaterally? Change the code?

**Comment by potiuk** [comment]
Well - it was you who proposed  to do something common between pydantic and sqlachemy so you expanded scope and discussions yourself, so yes it took longer and had more twists.

 I thinl up until then there was only one extra suggestion from TP so that you move things to sqlalchemy. If you need exact instructions here it is:

* do not make DRY code between Pydantic and sqlalchemy and pydantic at all. You can have separate PR for that f you want
* move the code to check sqlalchemy to utils.sqlalchemy

And yes, it happens and is expected. Sometimes small prs takes days or weeks and you will get different twists along the way and people looking at what you propose will have different opinions. And sometimes it get few iterations.

**Comment by Dev-iL** [comment]
As of the latest changes (submitted a few hours ago), both of the points you mentioned are implemented. Is it acceptable to resolve the threads then?

File: airflow/settings.py
Comment (Dev-iL) [reply airflow/settings.py]: Done.

**Comment by Dev-iL** [comment]
...merge?

**Review by potiuk — APPROVED** [review]

**Pull request merged**
